### PR TITLE
Sticky sessions implementation

### LINKED ISSUE #593: Sticky sessions scheduler
Depends on #235.

Actually foundation for [Sticky cookie](https://github.com/tempesta-tech/tempesta/issues/29) is application layer client identification, i.e. `hash(client IP, User-Agent)` identifies a 'virtual' client ('virtual' because there could be different clients behind the same proxy as well as couple of different browsers on the same device). 

The identifier is used for scheduling the client requests to appropriate server (see [Nginx sticky module](https://github.com/lusis/nginx-sticky-module) and [HAProxy sticky sessions](http://blog.haproxy.com/2012/03/29/load-balancing-affinity-persistence-sticky-sessions-what-you-need-to-know/)) or for security issues. The second use case (HMAC key in particular) is the subject for #235. While the first application, sticky sessions scheduling, wasn't developed at all.

So hash scheduler must be extended such that it can use Tempesta sticky cookie value to decide which upstream server the message should be scheduled to. This will guarantee that HTTP session will be always processed by the same upstream server. No one current scheduler can guarantee HTTP sessions persistence, so Tempesta can break user experience.

Overall requests scheduling must have important property: if Tempesta receives first request within a new HTTP session, it must schedule the request, and correspondingly the whole session, to upstream using some other scheduling algorithm like round-robin or one of [advanced load balancing algorithms](https://github.com/tempesta-tech/tempesta/issues/565). Actually hash scheduler using URI and Host values is also suitable for initial scheduling. See comment in `tfw_http_req_cache_cb()`.

#235 introduces `TfwHttpReq->TfwSession->TfwConnection:srv_conn`. So if `req->sess == NULL` in `tfw_http_req_cache_cb()`, then `tfw_sched_get_srv_conn()` is called. Otherwise the function should check it whether it's alive first and just get the connection if so (likely). Otherwise it must get a new connection for server identified by `req->sess->srv_conn->peer`. The same scheduler, which originally scheduled the session must be used to get a new connection for the same server. To implement this logic lets rename `sched_srv` callback to `sched_sg_conn` and introduce the new callback `TfwConnection *(*sched_srv_conn)(TfwMsg *msg, TfwServer *srv)` which scheduled connection only from `srv` argument or return error if all the connections are dead. If `sched_srv_conn()` fails, then we should do one of two configurable options: print error log message and return 50x code to client or print warning error message and fallback to `tfw_sched_get_srv_conn()`. To implement `sched_srv_conn()` RR scheduler needs do very similar things as `tfw_sched_rr_get_srv_conn()`: find the server by pointer in the first loop and get a connection from it by the same residual. Hash scheduler can just ignore differ server in the loop like `tfw_sched_hash_get_srv_conn()`.

Also don't forget to update README.

Also please implement appropriate unit test as for other schedulers.

### PR DESCRIPTION
**Pull request by ikoveshnikov**
Change log:

* fix #593 implement sticky sessions scheduler
* fix setting unlimited sticky cookie life time
* improve stability of unit tests

comment https://github.com/tempesta-tech/tempesta/issues/76#issuecomment-287572183 not implemented in the patch, instead i work on it in separate branch since some discussion may be needed.

### CODE REVIEW

File: tempesta_fw/http_sess.c [hunk 1]
```diff
@@ -720,6 +725,18 @@ tfw_http_sticky_secret_cfg(TfwCfgSpec *cs, TfwCfgEntry *ce)
 	return 0;
 }
 
+static int
+tfw_http_sticky_sess_lifetime_cfg(TfwCfgSpec *cs, TfwCfgEntry *ce)
+{
+	int r = tfw_cfg_set_int(cs, ce);
+
+	/* @sess_lifetime value of 0 means unlimited. */
+	if (!r && !tfw_cfg_sticky.sess_lifetime)
```
Comment (ikoveshnikov) [inline tempesta_fw/http_sess.c]: `0` in terms of `sess_lifetime` means `unlimited`, but setting actually zero to the parameter makes session to be removed right after it was created. As discussed before, `UINT_MAX` value should represent `unlimited` value for the code, but for the user it should remain as `0`.

I can rewrite condition as follows if you find that more meaningful:
```
if (!r && !*(cs->dest))
    *(cs->dest) = UINT_MAX;
```

File: tempesta_fw/http_sess.c [hunk 2]
```diff
@@ -736,12 +753,16 @@ TfwCfgMod tfw_http_sess_cfg_mod = {
 			.allow_none = true,
 		},
 		{
+			/* Value is parsed as int, set max to INT_MAX*/
 			.name = "sess_lifetime",
 			.deflt = "0",
-			.handler = tfw_cfg_set_int,
+			.handler = tfw_http_sticky_sess_lifetime_cfg,
 			.dest = &tfw_cfg_sticky.sess_lifetime,
```
Comment (ikoveshnikov) [inline tempesta_fw/http_sess.c]: In `tfw_http_sticky_sess_lifetime_cfg` i use `tfw_cfg_set_int` for parsing int value and check it for min-max borders. To use that i need to set `.dest` either here or in `tfw_http_sticky_sess_lifetime_cfg`.

File: tempesta_fw/http_sess.c (hunk 1 above)
Comment (keshonok) [inline tempesta_fw/http_sess.c]: Yes, that would make this function to NOT depend on the specific variable.
That would also make it generic for getting a number value for a directive, and setting is to max value if the number is zero (also a default value).

File: tempesta_fw/http_sess.c [hunk 3]
```diff
@@ -83,10 +84,10 @@ static TfwCfgSticky tfw_cfg_sticky;
...
 static char tfw_sticky_key[STICKY_KEY_MAXLEN];
 
-SessHashBucket sess_hash[SESS_HASH_SZ] = {
-	[0 ... (SESS_HASH_SZ - 1)] = {
-		HLIST_HEAD_INIT,
-	}
+static SessHashBucket sess_hash[SESS_HASH_SZ] = {
+//	[0 ... (SESS_HASH_SZ - 1)] = {
+//		HLIST_HEAD_INIT,
+//	}
```
Comment (krizhanovsky) [inline tempesta_fw/http_sess.c]: Don't use C++ style comments. Don't leave dead (just commented out) code. Why did you remove the initialization? I know that it's just zeroying, but using the initialization is still a good practice. At least use the list initializer in `tfw_http_sess_init()` near to spinlock initialization.

File: tempesta_fw/http_sess.c [hunk 4]
```diff
@@ -565,7 +566,7 @@ tfw_http_sess_obtain(TfwHttpReq *req)
 			goto found;
 	}
 
-	if (!(sess = kmem_cache_alloc(sess_cache, GFP_ATOMIC))) {
+	if (!(sess = kmem_cache_zalloc(sess_cache, GFP_ATOMIC))) {
```
Comment (krizhanovsky) [inline tempesta_fw/http_sess.c]: Don't use __GFP_ZERO w/o need, that's relatively expensive thing to make the kernel write zeroes to free pages. Instead, you can just write 3 NULLs  to the new pointer fields.

File: tempesta_fw/http_sess.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ */
+struct tfw_http_sess_t {
+	unsigned char		hmac[SHA1_DIGEST_SIZE];
+	struct hlist_node	hentry;
+	atomic_t		users;
+	unsigned long		ts;
+	unsigned long		expires;
+	TfwStickyConn		st_conn;
+};
+
```
Comment (krizhanovsky) [inline tempesta_fw/http_sess.h]: Extra empty line. Why do you define TfwHttpSess ouside of the file?

File: tempesta_fw/sched.c [hunk 6]
```diff
@@ -71,6 +70,166 @@ tfw_sched_get_srv_conn(TfwMsg *msg)
...
+ * Find an outgoing connection for client with tempesta sticky cookie.
+ * @sess is not null when calling the function.
+ *
+ * Reuse req->sess->st_conn.srv_conn if it is alive. If not,
+ * then get a new connection for req->sess->srv_conn->peer from
+ * an appropriate scheduler. That eliminates the long generic
+ * scheduling work flow.
+ */
+static inline TfwSrvConn *
+__get_sticky_srv_conn(TfwMsg *msg, TfwHttpSess *sess)
```
Comment (krizhanovsky) [inline tempesta_fw/sched.c]: Basically, `sched.c` isn't an appropriate place for Sticky cookie scheduling. `sched.c` is the generic scheduling logic. `__get_sticky_srv_conn()`, `__try_conn()` and maybe some other functions are good candidates for http_sess.c or a new file `sched/tfw_sched_sticky.c`. Also keep in mind #685 which is going to extend the logic even more.

File: tempesta_fw/sched.c [hunk 7]
```diff
@@ -71,6 +70,166 @@ tfw_sched_get_srv_conn(TfwMsg *msg)
...
+{
+	TfwStickyConn *st_conn = &sess->st_conn;
+	TfwSrvConn *srv_conn;
+
+	read_lock(&st_conn->lock);
```
Comment (krizhanovsky) [inline tempesta_fw/sched.c]: Why do you use read lock for `__try_conn()` which basically calls lock-free routines only?

File: tempesta_fw/sched.c [hunk 8]
```diff
@@ -71,6 +70,166 @@ tfw_sched_get_srv_conn(TfwMsg *msg)
...
+		srv_conn = backup_sg->sched->sched_sg_conn(msg, backup_sg);
+	}
+
+	if (unlikely(!srv_conn))
+		TFW_DBG2("sched: Unable to select server from group '%s'\n",
+			 sg->name);
+
+	return srv_conn;
+}
+EXPORT_SYMBOL(tfw_sched_get_sg_srv_conn);
```
Comment (krizhanovsky) [inline tempesta_fw/sched.c]: Please define the function before the first usage of it, i.e. before `__get_sticky_srv_conn()`.

File: tempesta_fw/sched/tfw_sched_hash.c [hunk 9]
```diff
@@ -45,22 +45,25 @@
...
 
 typedef struct {
-	TfwSrvConn	*srv_conn;
-	unsigned long	hash;
-} TfwConnHash;
+	size_t			conn_n;
+	TfwServer		*srv;
+	TfwSrvConn		*conn[TFW_SRV_MAX_CONN];
+	unsigned long		hash[TFW_SRV_MAX_CONN];
+} TfwHashSrv;
```
Comment (krizhanovsky) [inline tempesta_fw/sched/tfw_sched_hash.c]: Comment https://github.com/tempesta-tech/tempesta/pull/666/files#r94275356 is still not addressed. I saw your PR https://github.com/tempesta-tech/tempesta/pull/714/files, but it's somehow already closed.

File: tempesta_fw/http_sess.c (hunk 3 above)
Comment (ikoveshnikov) [inline tempesta_fw/http_sess.c]: Didn't want C++ comments to appear here. I use IDE with integrated clang code analysis and clang gets stuck on this initialization and the whole IDE hangs every time i make changes to this file.  I will be twice more careful on sending patches next time.

File: tempesta_fw/sched.c (hunk 7 above)
Comment (ikoveshnikov) [inline tempesta_fw/sched.c]: If session is under failovering, connection stored in sessions private data is no more valid. Even if original server is suddenly back online.

Comment (krizhanovsky) [inline tempesta_fw/sched.c]: Isn't it better to use reference counting to make the session private data valid even under failovering?

File: tempesta_fw/http_sess.h (hunk 5 above)
Comment (ikoveshnikov) [inline tempesta_fw/http_sess.h]: In last PR i was advised by @keshonok  to move all the session related definitions from `http.h` to separate header. But  `TfwHttpSess` is required for `TfwHttpReq` in `http.h` and most functions in this header require definitions from `http.h`, so circular dependency take place here.

As for me, advice was very valuable: http.h is too big and placing there more and more definitions only breaks it to loosely connected parts.

**Pull request merged**